commandapi-core/src/main/java/dev/jorel/commandapi/executors/CommandArguments.java

```diff
@@ -443,4 +457,153 @@ public <T> Optional<T> getOptionalUnchecked(String nodeName) {
+	 * <p>
+	 * This attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param nodeName The node name of the argument
+	 * @param argumentType The class that represents the argument
+	 * @return An <code>Optional</code> holding the argument, or an empty <code>Optional</code> if the argument was not found.
+	 */
+	public <T> Optional<T> getOptionalByClass(String nodeName, Class<T> argumentType) {
+		T argument = getByClass(nodeName, argumentType);
+		return Optional.ofNullable(argument);
+	}
+
+	/**
+	 * Returns an argument based on its index. This also attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param index The index of the argument
+	 * @param argumentType The class that represents the argument
+	 * @return The argument at the given index, or null if the argument was not found.
+	 */
+	@Nullable
+	public <T> T getByClass(int index, Class<T> argumentType) {
+		Object argument = get(index);
+		return castArgument(argument, argumentType, index);
+	}
+
+	/**
+	 * Returns an argument based on its index or a default value if the argument wasn't found.
+	 * <p>
+	 * If the argument was found, this method attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param index The index of the argument
+	 * @param argumentType The class that represents the argument
+	 * @param defaultValue The default value to return if the argument wasn't found
+	 * @return The argument at the given index, or the default value if the argument was not found.
+	 */
+	public <T> T getByClassOrDefault(int index, Class<T> argumentType, T defaultValue) {
+		T argument = getByClass(index, argumentType);
+		return (argument != null) ? argument : defaultValue;
+	}
+
+	/**
+	 * Returns an <code>Optional</code> holding the argument at the given index. This <code>Optional</code> can be empty if the argument was not given when running the command.
+	 * <p>
+	 * This attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param index The index of the argument
+	 * @param argumentType The class that represents the argument
+	 * @return An <code>Optional</code> holding the argument, or an empty <code>Optional</code> if the argument was not found.
+	 */
+	public <T> Optional<T> getOptionalByClass(int index, Class<T> argumentType) {
+		T argument = getByClass(index, argumentType);
+		return Optional.ofNullable(argument);
+	}
+
+	private <T> T castArgument(Object argument, Class<T> argumentType, Object argumentNameOrIndex) {
+		if (argument == null) {
+			return null;
+		}
+		if (!PRIMITIVE_TO_WRAPPER.getOrDefault(argumentType, argumentType).isAssignableFrom(argument.getClass())) {
+			throw new IllegalArgumentException(buildExceptionMessage(argumentNameOrIndex, argument.getClass().getSimpleName(), argumentType.getSimpleName()));
+		}
```

Comment:
I don't get what's going on here. Say I'm using a command with a single `AngleArgument` argument (which returns a `float` primitive).

Say we use the following to get the first argument:
```java
getByClass(0, float.class);
```

Using `PRIMITIVE_TO_WRAPPER` gives us:
```java
!PRIMITIVE_TO_WRAPPER.getOrDefault(float.class, float.class).isAssignableFrom(float.class)
```
which evaluates to the following:
```java
!Float.class.isAssignableFrom(float.class)
```
which evaluates to:
```java
!false
```
So we always throw an exception here? Or is my static-code analysis deceiving me and I'm missing something very obvious here?

Reply:
Yeah, this is one of first things I tested.
I think what's going on here is that because we store every argument as an `Object`, the primitive types are being boxed which is why we're not comparing to the primitive types.
So the `AngleArgument` becomes something like this:
```java
getByClass(0, float.class);
```
Then the primitive wrapper thing:
```java
!PRIMITIVE_TO_WRAPPER.getOrDefault(float.class, float.class).isAssignableFrom(Float.class)
```

This evaluates to:
```java
!Float.class.isAssignableFrom(Float.class)
```

This finally evaluates to:
```java
!true
```